ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParsePayload.java

```diff
@@ -75,132 +76,133 @@ public static ParsePayload of(String schemasLocation) {
   private transient PipelineMetadataStore metadataStore;
   private transient CRC32 crc32;
 
+  final TupleTag<PubsubMessage> outputTag = new TupleTag<>() {
+  };
+  final TupleTag<PubsubMessage> failureTag = new TupleTag<>() {
+  };
+
   private ParsePayload(String schemasLocation) {
     this.schemasLocation = schemasLocation;
   }
 
   @Override
   public Result<PCollection<PubsubMessage>, PubsubMessage> expand(
       PCollection<PubsubMessage> messages) {
-    return messages.apply(FlatMapElements.into(TypeDescriptor.of(PubsubMessage.class)) //
-        .via(new Fn()) //
-        .exceptionsInto(TypeDescriptor.of(PubsubMessage.class)) //
-        .exceptionsVia((WithFailures.ExceptionElement<PubsubMessage> ee) -> {
-          try {
-            throw ee.exception();
-          } catch (IOException | SchemaNotFoundException | ValidationException
-              | MessageScrubberException | DeprecatedMessageException e) {
-            return FailureMessage.of(ParsePayload.class.getSimpleName(), //
-                ee.element(), //
-                ee.exception());
-          }
-        }));
+    PCollectionTuple parsed = messages
+        .apply(ParDo.of(new Fn()).withOutputTags(outputTag, TupleTagList.of(failureTag)));
+    return WithFailures.Result.of(parsed.get(outputTag), parsed.get(failureTag));
```

Comment:
Keeping `WithFailures.Result` here instead of plain `PCollectionTuple` as the former makes handling errors upstream easier.